pkg/spdk/engine.go

```diff
@@ -1128,42 +1060,124 @@ func (e *Engine) ReplicaShallowCopy(dstReplicaName, dstReplicaAddress string) (e
 		prevSnapshotName = currentSnapshotName
 	}
 
-	// TODO: The rebuilding lvol of the dst replica is actually the head. Need to make sure the head stands behind to the correct snapshot.
-	//  Once we start to use a separate rebuilding lvol rather than the head, we can remove the below code.
-	if !rpcSrcReplica.Snapshots[prevSnapshotName].Children[types.VolumeHead] {
-		for snapshotName, snapApiLvol := range rpcSrcReplica.Snapshots {
-			if !snapApiLvol.Children[types.VolumeHead] {
-				continue
-			}
-			if err = srcReplicaServiceCli.ReplicaRebuildingSrcDetach(srcReplicaName, dstReplicaName); err != nil {
-				return err
+	e.log.Infof("Engine shallow copied all snapshots from rebuilding src replica %s to rebuilding dst replica %s", srcReplicaName, dstReplicaName)
+
+	return nil
+}
+
+// replicaAddFinish tries its best to finish the replica add no matter if the dst replica is rebuilt successfully or not.
+// It returns fatal errors that lead to engine unavailable only. As for the errors during replica rebuilding wrap-up, it will be logged and ignored.
+func (e *Engine) replicaAddFinish(srcReplicaServiceCli, dstReplicaServiceCli *client.SPDKClient, srcReplicaName, dstReplicaName string) (err error) {
+	updateRequired := false
+
+	e.Lock()
+	defer func() {
+		e.Unlock()
+
+		if updateRequired {
+			e.UpdateCh <- nil
+		}
+	}()
+
+	defer func() {
+		if err != nil {
+			if e.State != types.InstanceStateError {
+				e.State = types.InstanceStateError
+				updateRequired = true
 			}
-			if err = dstReplicaServiceCli.ReplicaRebuildingDstSnapshotRevert(dstReplicaName, snapshotName); err != nil {
-				return err
+			e.ErrorMsg = err.Error()
+		} else {
+			if e.State != types.InstanceStateError {
+				e.ErrorMsg = ""
 			}
-			break
 		}
+	}()
+
+	// Blindly ask the source replica to detach the rebuilding lvol
+	if srcReplicaErr := srcReplicaServiceCli.ReplicaRebuildingSrcDetach(srcReplicaName, dstReplicaName); srcReplicaErr != nil {
+		e.log.WithError(srcReplicaErr).Errorf("Engine failed to detach the rebuilding lvol for rebuilding src replica %s, will ignore this error and continue", srcReplicaName)
 	}
```

Comment:
Will there be any orphaned attached rebuilding lvol if failed to detach?

Reply:
Theoretically YES. But we have not idea to handle it if this detachment fails. And we cannot make it block the flow

Reply:
I see. I think we can figure out a solution to it.
Can you leave a comment for it? Thank you.